lib/galaxy/security/passwords.py

```diff
@@ -67,29 +67,38 @@ def check_password_PBKDF2(guess, hashed):
 _pack_int = Struct('>I').pack
 
 
-def pbkdf2_bin(data, salt, iterations=1000, keylen=24, hashfunc=None):
+def pbkdf2_bin(data, salt, iterations=COST_FACTOR, keylen=KEY_LENGTH, hashfunc=HASH_FUNCTION):
     """Returns a binary digest for the PBKDF2 hash algorithm of `data`
     with the given `salt`.  It iterates `iterations` time and produces a
-    key of `keylen` bytes.  By default SHA-1 is used as hash function,
+    key of `keylen` bytes.  By default SHA-256 is used as hash function,
     a different hashlib `hashfunc` can be provided.
     """
-    hashfunc = hashfunc or hashlib.sha1
-    mac = hmac.new(smart_str(data), None, hashfunc)
-
-    def _pseudorandom(x, mac=mac):
-        h = mac.copy()
-        h.update(x)
-        digest = h.digest()
-        if six.PY2:
-            return digest, [ord(_) for _ in digest]
-        return digest, digest
-
-    buf = []
-    salt = smart_str(salt)
-    for block in range(1, -(-keylen // mac.digest_size) + 1):
-        digest, rv = _pseudorandom(salt + _pack_int(block))
-        for _ in range(iterations - 1):
-            digest, u = _pseudorandom(digest)
-            rv = starmap(xor, zip(rv, u))
-        buf.extend(rv)
-    return bytes(bytearray(buf))[:keylen]
+    if hasattr(hashlib, 'pbkdf2_hmac'):
+        # Use hashlib.pbkdf2_hmac, new in python 2.7.8
+        data = smart_str(data)
+        salt = smart_str(salt)
+        return hashlib.pbkdf2_hmac(hashfunc, data, salt, iterations, keylen)
+    else:
+        # Use the old implementation of hashing as a fallback
+        if isinstance(hashfunc, six.string_types):
+            hashfunc = getattr(hashlib, hashfunc, None)
+        hashfunc = hashfunc or hashlib.sha1
+        mac = hmac.new(smart_str(data), None, hashfunc)
+
+        def _pseudorandom(x, mac=mac):
+            h = mac.copy()
+            h.update(x)
+            digest = h.digest()
+            if six.PY2:
```

Comment:
I guess we don't need the guard anymore

Reply:
Ahh, good point -- I left the whole block intact for the backport, but this can be simplified.  I guess technically python 3.3 doesn't have hashlib.pbkdf2_hmac, either, but surely nobody made the leap to 3 and is running something that old that we don't support anyway.  I'll drop this.